jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java

```diff
@@ -0,0 +1,291 @@
+/*
+ * (C) Copyright 2018-2018, by Ira Justus Fesefeldt and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.decomposition;
+
+import java.util.*;
+
+import org.jgrapht.*;
+import org.jgrapht.alg.util.*;
+import org.jgrapht.graph.*;
+
+/**
+ * An abstract builder class for nice tree decompositions, which builds the tree decomposition in a
+ * top-down manner.
+ * <p>
+* A tree-decomposition of a graph $G = (V, E)$ is a pair $(X, T) = (\{X_i\ |\ i\in I\}, (I, F))$ where
+ * $X = \{X_i\ |\ i \in I \}$ is a family of subsets of $V$, and $T = (I, F)$ is a tree, such that
+ * <ul>
+ * <li>Union of the sets $X_i$ equals to $V$</li>
+ * <li>for every edge $e = (u,v)$ there exists a set $X_i$ such that $u \in X_i$ and $v \in X_i$</li>
+ * <li>if both $X_i$ and $X_j$ contain a vertex $v$ then every set $X_k$ on the simple path from $X_i$ to $X_j$
+ * contains vertex $v$.</li>
+ * </ul>
+ * <br>
+ * A nice tree decomposition is a special tree decomposition, which satisfies the properties:
+ * <ul>
+ * <li>for root $r \in V(T)$ and leaf $l \in V(T): |b(r)|=|b(t)|=1$</li>
+ * <li>every non-leaf node $t \in V(T)$ is of one of the following three types:
+ * <ul>
+ * <li>forget node: $t$ has exactly one child $d$ and $b(t) \cup \{ w\} = b(d)$ for some $w \in
+ * V(G)\setminus b(t)$</li>
+ * <li>introduce node: $t$ has exactly one child $d$ and $b(t) = b(d) \cup \{ w\}$ for some $w \in
+ * V(G)\setminus b(d)$</li>
+ * <li>join node: $t$ has exactly two children $d_1$, $d_2$ and $b(t)=b(d_1)=b(d_2)$</li>
+ * </ul>
+ * </ul>
+ * <p>
+ * See:<br>
+ * Bodlaender, Hans &amp; Kloks, Ton. (1991). Better Algorithms for the Pathwidth and Treewidth of
+ * Graphs. 544-555. 10.1007/3-540-54233-7_162.<br>
+ * for a more complete description of tree decomposition and nice tree decomposition.
+ * 
+ * @author Ira Justus Fesefeldt (PhoenixIra)
+ *
+ * @param <V> the vertices of the graph
+ */
+abstract public class NiceDecompositionBuilder<V>
+{
+
+    // resulting decomposition
+    protected Graph<Integer, DefaultEdge> decomposition;
```

Comment:
I am wondering why not use directly something like `Graph<Set<V>, DefaultEdge>` for the decomposition (and for the result of the decomposition).

Reply:
This was also my first attempt, but a decomposition (especially a nice decomposition) can multiple nodes with same sets. Just look at the definition of a join node, where the join node and both of their children have the same set. But the graph class (or more precise: the vertex set class) would consider these three nodes the same. So my two options was to either use a Set with modified equal/hashvalue getter, a wrapper class or a Map from Node to Sets. Since the last one is equal to the formal definition, I opt for this method.